media/windows/ECC-Ecobee/ECC Pycobee Data v01.00.py

```diff
@@ -0,0 +1,1769 @@
+# the historical 5-minute interval runtime data for each thermostat.  thermRuntime stores the snapshot data
+# obtained through the thermostat details API call.
+#       Temporary table definitions to limit data retrieval until lists parsing added:
+SQLite_table_dict = {'settings': 'thermSettings', 'location': 'thermLocation',
+                     'house_details': 'thermHouseDetails', 'version': 'thermVersion'}
+'''
+SQLite_table_dict = {'settings': 'thermSettings', 'runtime': 'thermRuntime',
+                     'extended_runtime': 'thermExtendedRuntime', 'electricity': 'thermElectricity',
+                     'location': 'thermLocation', 'technician': 'thermTechnician',
+                     'utility': 'thermUtility', 'management': 'thermManagement', 'weather': 'thermWeather',
+                     'program': 'thermProgram', 'house_details': 'thermHouseDetails',
+                     'oem_cfg': 'thermOemCfg', 'notification_settings': 'thermNotificationSettings',
+                     'privacy': 'thermPrivacy', 'version': 'thermVersion',
+                     'security_settings': 'thermSecuritySettings'}
+'''
+
+def main():
+    # Global variables used for informational logging
+    global dup_update_cnt_total
+    global dup_update_cnt_this_thermostat
+    global blank_rec_cnt_total
+    global blank_rec_cnt_this_thermostat
+    global db_table_recs_written
+
+    dup_update_cnt_total = 0
+    dup_update_cnt_this_thermostat = 0
+    blank_rec_cnt_total = 0
+    blank_rec_cnt_this_thermostat = 0
+    snapshot_recs_written_total = 0
+    db_table_recs_written = {}  # Initialize a dictionary for storing recs written by table for snapshots
+
+    print(F"*** Initiating Ecobee thermostat data retrieval ***")
+    logger.info(F"*** Initiating Ecobee thermostat data retrieval ***")
+    print(F"*** ECC Ecobee data retrieval version {eccpycobee_version}, {eccpycobee_date} ***")
+    logger.info(F"*** ECC Ecobee data retrieval version {eccpycobee_version}, {eccpycobee_date} ***")
+
+    # Attempt to open the credentials / authorization file and read contents
+    try:
+        pyecobee_db = shelve.open(ECCAuthorize,  # read the stored authorization file
+                                  writeback=True,
+                                  protocol=4)  # protocol 4 is the latest pickle version, > Python 3.4
+        logger.debug(F"Shelve data structure:  Keys: {len(pyecobee_db)}")
+        logger.debug(F"...elements:  {list(pyecobee_db)}")
+        auth_list = list(pyecobee_db)
+        for pelement in auth_list:
+            logger.debug(F"...contents:  {pelement}: {pyecobee_db.get(pelement)}")
+
+        # Test to see if the read credentials are empty; if so, initialize the Ecobee service object.
+        # This would typically happen at first run when there are no stored credentials.
+        if len(pyecobee_db) == 0:
+            app_key = "Vy72kTcU63iWryQ9YG5T4faruTFKxIwz"  # default, if not read from shelve
```

Comment:
If by secret you’re referring to something like an API key, then yes.  It does contain the API key used for R/O access to the thermostat data.  I’ve considered multiple ways of further obfuscation (storing the value in a Keyring for instance), but if someone has access to the code, I don’t see that it makes much difference in the end.  I’m open to other suggestions though...

Reply:
In cases like this, I typically make secret values like this be loaded from an external file that is not maintained in github.

Given that this key has already been compromised (i.e., posted on github), it would probably be good to revoke this key, get a new key, and update the code to load the key from an external file.  Send the key to me some other way (e.g., google drive or somesuch).

Reply:
Yep, I knew about it already being compromised, but wasn’t overly concerned given it was R/O, and the need for an initial authorization to occur when the new tokens are requested.  I assumed we could deal with that too when you were ready to move it.  I’ll redirect the reading of the default key to any external file before I post the very latest code.  You’ll also want to have the latest database before running it in production, which I’ll move somewhere at the same time.

Reply:
Cool, thanks!